api/v1alpha1/gateway_types.go

```diff
@@ -54,78 +57,200 @@ type GatewayList struct {
+// names provided by the client at both the TLS and the HTTP protocol layers.
+type HostnameMatch struct {
+	// Match specifies how the host name provided by the client should be
+	// matched against the given value.
+	//
+	// +optional
+	// +kubebuilder:validation:Enum=Domain;Exact
+	// +kubebuilder:default=Exact
+	Match HostnameMatchType `json:"match" protobuf:"bytes,1,name=match"`
+
+	// Name contains the name to match against. This value must be a
+	// fully qualified host or domain name as defined by
+	// [RFC 3986](https://tools.ietf.org/html/rfc3986#section-3.2.2).
+	// Note the following deviations from the "host" part of the URI as defined in the RFC:
+	//
+	// 1. IPs are not allowed.
+	// 2. The `:` delimiter is not respected because ports are not allowed.
+	//
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	Name string `json:"value" protobuf:"bytes,2,name=value"`
+}
+
+// Listener embodies the concept of a logical endpoint where a
+// Gateway can accept network connections.
+//
+// The GatewayClass may collapse Listener definitions into single
+// implementation-defined acceptor configuration. This allows Gateways to
+// be implemented in terms of virtual services configured within the same
+// proxy server process.
+//
+// Listeners may be considered collapsible if all of the following
+// conditions are true:
+//
+// 1. the Protocol field is "HTTP", "HTTPS" or "TLS",
+// 2. the Hostname field is specified
+// 3. and the Port field matches
+//
+// As a special case, each group of collapsed Listeners may
+// contain up to one Listener that omits the Hostname field. Since
+// omitting the Hostname field captures all traffic directed to the
+// specified port, allowing more than one of these Listeners would
+// create either ambiguity or redundancy.
+//
+// Listeners that specify identical Hostname fields must not be
+// treated as collapsible. Put another way, each Hostname field in
+// a collapsible group must be unique in order to avoid ambiguity.
+//
+// If the GatewayClass collapses Listeners, the host name provided
+// in the incoming client request must be matched to a Listener to
+// find the correct set of Routes. The incoming host name must be matched
+// using the Hostname field for each listener in order of most to least
+// specific. That is, "Exact" matches must be processed before
+// "Domain" matches, which must be processed before Listeners that
+// omit the Hostname field.
+//
+// If the Gateway specifies multiple Listeners that have the same
+// Port value but are not collapsible, the GatewayClass must raise
+// a "PortConflict" condition on the Gateway.
 type Listener struct {
-	// Name is the listener's name and should be specified as an
-	// RFC 1035 DNS_LABEL [1]:
+	// Hostname specifies the virtual host name for protocol types
+	// that define this concept.
 	//
-	// [1] https://tools.ietf.org/html/rfc1035
+	// Incoming requests are matched against Hostname before processing
```

Comment:
I think we can put how it matches w/ the given types aboe.

Requests will be matched according to the given HostnameMatch type defined above...

Reply:
Rewritten a bit.